core/thread/src/TReentrantRWLock.cxx

```diff
@@ -221,100 +221,151 @@ void TReentrantRWLock<MutexT, RecurseCountsT>::WriteUnLock(TVirtualRWMutex::Hint
-void TReentrantRWLock<MutexT, RecurseCountsT>::Restore(std::unique_ptr<TVirtualMutex::State> &&state)
-{
-   TReentrantRWLockState<MutexT, RecurseCountsT> *pState = dynamic_cast<TReentrantRWLockState<MutexT, RecurseCountsT> *>(state.get());
-   if (!pState) {
-      if (state) {
-         SysError("Restore", "LOGIC ERROR - invalid state object!");
-         return;
-      }
-      // No state, do nothing.
+void TReentrantRWLock<MutexT, RecurseCountsT>::Apply(std::unique_ptr<StateDelta> &&state) {
+   if (!state) {
+      Error("TReentrantRWLock::Apply", "Cannot apply empty delta!");
       return;
    }
 
-   // At a restore point, this thread should not be holding any part
-   // of the lock (if it does the following code will forget about it)
-   assert( *(pState->fReadersCountLoc) == 0);
-   // assert( auto local = fRecurseCounts.GetLocal() && fRecurseCounts.IsNotCurrentWriter(local) )
+   using StateDelta_t = TReentrantRWLockStateDelta<MutexT, RecurseCountsT>;
+   const StateDelta_t* typedDelta = static_cast<const StateDelta_t*>(state.get());
 
-   const auto readerCount = pState->fReadersCount;
+   if (typedDelta->fDeltaWriteRecurse < 0) {
+      Error("TReentrantRWLock::Apply", "Negative write recurse count delta!");
+      return;
+   }
+   if (typedDelta->fDeltaReadersCount < 0) {
+      Error("TReentrantRWLock::Apply", "Negative read count delta!");
+      return;
+   }
+   AssertReadCountLocIsFromCurrentThread(typedDelta->fReadersCountLoc);
 
-   if (pState->fIsWriter) {
-      WriteLock();
-      // Now that we go the lock, fix up the recursion count.
-      std::unique_lock<MutexT> lock(fMutex);
-      fRecurseCounts.fWriteRecurse = pState->fWriteRecurse;
-      *(pState->fReadersCountLoc) = readerCount;
-      fReaders +=  readerCount;
-   } else if (readerCount) {
+   if (typedDelta->fDeltaReadersCount != 0) {
       ReadLock();
-      // Now that we got the read lock, fix up the local recursion count.
-
-      *(pState->fReadersCountLoc) = readerCount;
-      fReaders += readerCount - 1;
+      *typedDelta->fReadersCountLoc += typedDelta->fDeltaReadersCount - 1;
+   } else if (typedDelta->fDeltaWriteRecurse != 0) {
+      WriteLock();
+      fRecurseCounts.fWriteRecurse += typedDelta->fDeltaWriteRecurse - 1;
+      *typedDelta->fReadersCountLoc += typedDelta->fDeltaReadersCount;
    }
+}
 
-   // Do something.
-   //::Fatal("Restore()", "Not implemented, contact pcanal@fnal.gov");
+//////////////////////////////////////////////////////////////////////////
+/// Assert that presumedLocalReadersCount really matches the local read count.
+/// Print an error message if not.
+
+template <typename MutexT, typename RecurseCountsT>
+void TReentrantRWLock<MutexT, RecurseCountsT>::AssertReadCountLocIsFromCurrentThread(const size_t* presumedLocalReadersCount) const
+{
+   (void)presumedLocalReadersCount;
+#ifdef DEBUG
+   auto local = fRecurseCounts.GetLocal();
+   std::lock_guard<MutexT> lock(fMutex);
```

Comment:
Not essential since this is DEBUG code but the lifetime of the lock_guard can (should) be limited to just the statement containing the call to GetLocalReadersCount